Classes/Base/Locale.php

```diff
@@ -34,29 +34,35 @@ public static function get(\Aimeos\MShop\ContextIface $context,
     {
         if (!isset(self::$locale)) {
             $config = $context->config();
-
-
             $sitecode = $config->get('mshop/locale/site', 'default');
-            $name = $config->get('typo3/param/name/site', 'site');
-
-            if ($request !== null && $request->hasArgument($name) === true) {
-                $sitecode = $request->getArgument($name);
-            }
-
-
             $lang = $config->get('mshop/locale/language', '');
-            $name = $config->get('typo3/param/name/language', 'locale');
-
-            if ($request !== null && $request->hasArgument($name) === true) {
-                $lang = $request->getArgument($name);
-            }
-
-
             $currency = $config->get('mshop/locale/currency', '');
-            $name = $config->get('typo3/param/name/currency', 'currency');
 
-            if ($request !== null && $request->hasArgument($name) === true) {
-                $currency = $request->getArgument($name);
+            if ($request !== null) {
+                $name = $config->get('typo3/param/name/site', 'site');
+                $siteAttribute = $request->getAttribute('site');
+
+                if ($request->hasArgument($name) === true) {
+                    $sitecode = $request->getArgument($name);
+                } elseif ($siteAttribute && $siteAttribute->getSettings()->has('aimeosSite')) {
+                    $sitecode = $siteAttribute->getSettings()->get('aimeosSite');
+                }
+
+                $name = $config->get('typo3/param/name/language', 'locale');
+                $languageAttribute = $request->getAttribute('language');
+
+                if ($request->hasArgument($name) === true) {
+                    $lang = $request->getArgument($name);
+                } elseif ($languageAttribute) {
+                    $lang = substr($languageAttribute->getLocale(), 0, 5);
+                }
+
+                $name = $config->get('typo3/param/name/currency', 'currency');
+                if ($request->hasArgument($name) === true) {
+                    $currency = $request->getArgument($name);
+                } elseif ($languageAttribute && isset($languageAttribute->toArray()['aimeosCurrency'])) {
+                    $currency = $languageAttribute->toArray()['aimeosCurrency'];
+                }
```

Comment:
Wouldn't it be better to use the site attribute instead of the language attribute for the currency?
```
} else {
    $currency = $request->getAttribute('site')?->getSettings()->get('aimeos.currency');
}
```

Reply:
In many instances the language is also used as a country selector. And in those cases currency would make sense?!
I could change it so both work and the language overrides if set?

Reply:
OK, see your point and using the language object is enough:
```
$currency = $request->getAttribute('language')?->toArray()['currency'] ?? null;
```